src/Language/Fixpoint/Names.hs

```diff
@@ -58,195 +89,261 @@ import           Control.Applicative ((<$>))
 import           Data.Monoid (Monoid (..))
 #endif
 
+import           System.IO.Unsafe            (unsafePerformIO)
 import           Control.DeepSeq             (NFData (..))
 import           Control.Arrow               (second)
 import           Control.Monad               ((>=>))
-import           Data.Char                   (isAlpha, chr, ord)
+import           Data.IORef
+-- import           Data.Char                   (isAlpha, chr, ord)
 import           Data.Generics               (Data)
 import           Data.Hashable               (Hashable (..))
 import qualified Data.HashSet                as S
+import qualified Data.HashMap.Strict         as M
 import           Data.Interned               (intern, unintern)
 import           Data.Interned.Internal.Text
 import           Data.Maybe                  (fromMaybe)
-import           Data.String                 (IsString)
+import           Data.String                 (IsString(..))
 import qualified Data.Text                   as T
 import           Data.Binary                 (Binary (..))
 import           Data.Typeable               (Typeable)
 import           GHC.Generics                (Generic)
 
-import           Language.Fixpoint.Misc      (errorstar)
-
+import           Language.Fixpoint.Misc      (safeLookup, errorstar)
+import Debug.Trace
 
 ---------------------------------------------------------------
 -- | Symbols --------------------------------------------------
 ---------------------------------------------------------------
 
-symChars
-  =  ['a' .. 'z']
-  ++ ['A' .. 'Z']
-  ++ ['0' .. '9']
-  ++ ['_', '%', '.', '#']
-
 deriving instance Data     InternedText
 deriving instance Typeable InternedText
 deriving instance Generic  InternedText
 
-newtype Symbol = S InternedText
-                 deriving (Eq, Ord, Data, Typeable, Generic, IsString)
+{-@ type SafeText = {v: T.Text | IsSafe v} @-}
+type SafeText = T.Text
+
+-- | Invariant: a `SafeText` is made up of:
+--
+--     ['0'..'9'] ++ ['a'...'z'] ++ ['A'..'Z'] ++ '$'
+--
+--   If the original text has ANY other chars, it is represented as:
+--
+--     lq$i
+--
+--   where i is a unique integer (for each text)
```

Comment:
I **really** don't like this encoding scheme, I fear it will make the fq files (or god forbid the smt files) WAY harder to read.

I strongly suggest adopting a variant of GHC's [z-encoding](https://ghc.haskell.org/trac/ghc/wiki/Commentary/Compiler/SymbolNames#AlphanumericCharacters), so we can actually read the intermediate files.